### PR TITLE
[WIP] Fix issue #285 : save hive partitioned dataset using NativeExecutionEngine and DaskExecutionEngine

### PR DESCRIPTION
**Pull request by LaurentErreca**
Enable saving hive partitioned dataset using Native (pandas) or Dask execution engine. Work still in progress as test_take with Dask is failing.
Related to https://github.com/fugue-project/fugue/issues/285

### CODE REVIEW

**Comment by kvnkho** [comment]
I looked at the tests here and from what I see, the only failure is in the `duckdb` execution engine for the io tests that you added. Below is a log of the test. The `test_take` tests seem to be fine.

```
self = <tests.fugue_duckdb.test_execution_engine.DuckBuiltInTests testMethod=test_io>

    def test_io(self):
        path = os.path.join(self.tmpdir, "a")
        path2 = os.path.join(self.tmpdir, "b.test.csv")
        path3 = os.path.join(self.tmpdir, "c.partition")
        with self.dag() as dag:
            b = dag.df([[6, 1], [2, 7]], "c:int,a:long")
            b.partition(num=3).save(path, fmt="parquet", single=True)
            b.save(path2, header=True)
        assert FileSystem().isfile(path)
        with self.dag() as dag:
            a = dag.load(path, fmt="parquet", columns=["a", "c"])
            a.assert_eq(dag.df([[1, 6], [7, 2]], "a:long,c:int"))
            a = dag.load(path2, header=True, columns="c:int,a:long")
            a.assert_eq(dag.df([[6, 1], [2, 7]], "c:int,a:long"))
        with self.dag() as dag:
            b = dag.df([[6, 1], [2, 7]], "c:int,a:long")
            b.partition(by='c').save(path3, fmt="parquet", single=False)
            b.save(path2, header=True)
>       assert FileSystem().isdir(path3)
E       AssertionError
```

So the issue here appears to be that the added tests support partitioned files as output (which is what the PR is for), but duckdb fails the test because it [doesn't support partitioning yet](https://github.com/duckdb/duckdb/issues/2186). 

How do we proceed with this @goodwanghan ?

File: fugue_test/builtin_suite.py [hunk 1]
```diff
@@ -1185,6 +1186,19 @@ def test_io(self):
...
+                b = dag.df([[6, 1], [2, 7]], "c:int,a:long")
+                b.partition(by='c').save(path3, fmt="parquet", single=False)
+                b.save(path2, header=True)
+            assert FileSystem().isdir(path3)
+            assert FileSystem().isdir(os.path.join(path3, 'c=6'))
+            assert FileSystem().isdir(os.path.join(path3, 'c=2'))
+            # TODO: in test below, once issue #288 is fixed, use dag.load instead of pd.read_parquet
+            pd.testing.assert_frame_equal(
+                pd.read_parquet(path3).sort_values('a').reset_index(drop=True),
+                pd.DataFrame({'c': pd.Categorical([6, 2]), 'a': [1, 7]}).reset_index(drop=True),
```
Comment (goodwanghan) [inline fugue_test/builtin_suite.py:1199]: Should we assert c is `int` type?

**Comment by goodwanghan** [comment]
The failure on duckdb test is expected because duckdb itself does not have partitioning feature. But it is quite straightforward to convert duckdbdataframe to arrow table (`as_arrow`), and then save to parquet with partition, so we should also make that change, then this looks super nice.

**Comment by goodwanghan** [comment]
I apologize for the delay, my daughter was born in Feb, so I am extremely busy recently.

The PR looks great, we just need to change duckdb io function to pass the test suite.

You can see we have other work items to redesign the IO part for all engines. Because the current design is a bit messy. But please continue working on this PR and merge first. The code refactoring can happen later. As long as we have good unit tests, we can ship this feature first.

Thank you!

**Comment by LaurentErreca** [comment]
Hi! No problem at all with the delay.

Congrats for the recent birth of your daughter!

I'll try to fix the duckdb io function that is failing to pass the test.

I did some prototyping this week to try to fix issue #288 related to
reading of hive partitioned dataset with Native and Dask execution engine.
I think I'm close to a solution in native mode, but it seems more dificult
to deal with dask categorical type, as category values are set to string
type. So the inference of partition column type is more tricky.

Laurent

**Comment by goodwanghan** [comment]
@LaurentErreca do you plan to finalize this PR soon?

**Comment by LaurentErreca** [comment]
I've been working on this issue. For duckdb to pass the test, it must be able to write hive partitioned dataset. I propose to use arrow with param `partition_cols` instead of the duckdb write function which uses a `COPY TO` function under the hood. This function doesn't support hive partitioned write.

**Comment by goodwanghan** [comment]
> I've been working on this issue. For duckdb to pass the test, it must be able to write hive partitioned dataset. I propose to use arrow with param `partition_cols` instead of the duckdb write function which uses a `COPY TO` function under the hood. This function doesn't support hive partitioned write.

Yes exactly, we should just convert it to arrow and save (only if this partition key is specified). It is straightforward to convert duckdbdataframe to arrow table (`as_arrow`), and then save to parquet with partition.

**Comment by goodwanghan** [comment]
It looks good you only have a linting issue left @LaurentErreca

**Comment by LaurentErreca** [comment]
> > I've been working on this issue. For duckdb to pass the test, it must be able to write hive partitioned dataset. I propose to use arrow with param `partition_cols` instead of the duckdb write function which uses a `COPY TO` function under the hood. This function doesn't support hive partitioned write.
> 
> Yes exactly, we should just convert it to arrow and save (only if this partition key is specified). It is straightforward to convert duckdbdataframe to arrow table (`as_arrow`), and then save to parquet with partition.

This is done by replacing:
`        
if p.file_format not in self._format_save:
    self._fs.makedirs(os.path.dirname(uri), recreate=True)
    ldf = ArrowDataFrame(df.native.arrow())
`
with:
`
if (p.file_format not in self._format_save) or ("partition_cols" in kwargs):
    self._fs.makedirs(os.path.dirname(uri), recreate=True)
    ldf = ArrowDataFrame(df.native.arrow())
`

**Comment by goodwanghan** [comment]
> > > I've been working on this issue. For duckdb to pass the test, it must be able to write hive partitioned dataset. I propose to use arrow with param `partition_cols` instead of the duckdb write function which uses a `COPY TO` function under the hood. This function doesn't support hive partitioned write.
> > 
> > 
> > Yes exactly, we should just convert it to arrow and save (only if this partition key is specified). It is straightforward to convert duckdbdataframe to arrow table (`as_arrow`), and then save to parquet with partition.
> 
> This is done by replacing: ` if p.file_format not in self._format_save: self._fs.makedirs(os.path.dirname(uri), recreate=True) ldf = ArrowDataFrame(df.native.arrow())` with: `if (p.file_format not in self._format_save) or ("partition_cols" in kwargs): self._fs.makedirs(os.path.dirname(uri), recreate=True) ldf = ArrowDataFrame(df.native.arrow())`

I think you should use `df.as_arrow()` instead of `df.native.arrow()` because as_arrow is working on any dataframe. df.native assumes it is a duckdb relation, sometimes it is not.

File: fugue_duckdb/_io.py [hunk 2]
```diff
@@ -70,7 +70,7 @@ def save_df(
             NotImplementedError(f"{mode} is not supported"),
         )
         p = FileParser(uri, format_hint).assert_no_glob()
-        if p.file_format not in self._format_save:
+        if (p.file_format not in self._format_save) or ("partition_cols" in kwargs):
             self._fs.makedirs(os.path.dirname(uri), recreate=True)
             ldf = ArrowDataFrame(df.native.arrow())
```
Comment (goodwanghan) [inline fugue_duckdb/_io.py:75]: Please use `ArrowDataFrame(df.as_arrow())` instead

Comment (LaurentErreca) [reply fugue_duckdb/_io.py:75]: Ok

**Review by goodwanghan — APPROVED** [review]

**Pull request merged**